### PR TITLE
[feat] Add ability to track and visualize 3D figures

### PR DESCRIPTION
**Pull request by pabhermoso**
First version of a 3D Geometry Visualiser under Run Details. Please feel free to contact me and we can go over it.

I am currently struggling with the API Request under: aim/aim/web/ui/src/pages/RunDetail/GeometriesVisualizer/GeometriesVisualizer.tsx

I believe it gets stuck in "Pending".
Screenshot 2023-04-17 at 21 31 43

I created a Geometry Aim Object. To run this I created a Python environment and installed aim from source and then:
_cd aim/web/ui
npm install --legacy-peer-deps
npm start_

Could you please try to run it from your side and try to identify why the API request is not being completed?

That would be super appreciated!!!

### CODE REVIEW

**Comment by pabhermoso** [comment]
Maybe just one additional point for completeness. In the notebook insert_runs.ipynb I have an example of creating a Run with a geometry.
Screenshot 2023-04-18 at 14 04 28

**Comment by roubkar** [comment]
Thanks for the details @pabhermoso 
We will keep you updated with the progress on fixing blobs data fetching issue.

**Comment by pabhermoso** [comment]
Sorry, I just updated it now.

**Comment by roubkar** [comment]
Resolves https://github.com/aimhubio/aim/issues/2644

File: aim/sdk/sequences/geometry_sequence.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+from aim.sdk.objects.geometry import Geometry
+
+
+class Geometries(MediaSequenceBase):
+    """Class representing series of Geometry objects or Geometry lists."""
+
+    @classmethod
+    def allowed_dtypes(cls) -> Union[str, Tuple[str, ...]]:
+        geometry_typename = Geometry.get_typename()
+        return geometry_typename, f'list({geometry_typename})'
```
Comment (alberttorosyan) [inline aim/sdk/sequences/geometry_sequence.py:13]: Lets allow tracking of single geometry objects to improve UX when tracking and reading sequences.

File: aim/web/api/runs/pydantic_models.py [hunk 2]
```diff
@@ -221,6 +221,11 @@ class DistributionInfo(BaseModel):
 class FigureInfo(BaseModel):
     blob_uri: str
 
+class GeometryInfo(BaseModel):
+    caption: str
+    blob_uri: str
+    index: int
```
Comment (alberttorosyan) [inline aim/web/api/runs/pydantic_models.py:227]: Remove `index`, as only single Geometry objects should be tracked.

File: aim/web/api/runs/pydantic_models.py [hunk 3]
```diff
@@ -229,3 +234,4 @@ class NoteIn(BaseModel):
 ImageList = List[ImageInfo]
 TextList = List[TextInfo]
 AudioList = List[AudioInfo]
+GeometryList = List[GeometryInfo]
```
Comment (alberttorosyan) [inline aim/web/api/runs/pydantic_models.py:237]: No need for `GeometryList`

File: aim/web/api/runs/object_views.py [hunk 4]
```diff
@@ -182,3 +184,8 @@ class FigureApiConfig(CustomObjectApiConfig):
     resolve_blobs = True
     dump_record_fn = lambda x: x.data  # noqa E731
     model = FigureInfo
+
+class GeometryApiConfig(CustomObjectApiConfig):
+    sequence_type = Geometries
+    resolve_blobs = False
+    model = GeometryList
```
Comment (alberttorosyan) [inline aim/web/api/runs/object_views.py:191]: `model = GeometryInfo`

File: aim/sdk/objects/geometry.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+    def _prepare(self, data, **extra) -> None:
+        assert isinstance(data, bytes)
+
+        for k, v in extra.items():
+            self.storage[k] = v
+        self.storage['data'] = BLOB(data=data)
+
+    # def to_numpy(self):
```
Comment (alberttorosyan) [inline aim/sdk/objects/geometry.py:84]: please remove the commented method

File: aim/sdk/objects/geometry.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    STL = 'stl'
+    OBJ = 'obj'
+
+    geometry_formats = (STL, OBJ)
+
+    def __init__(self, data, format: str = '', caption: str = '', rate: int = None):
+        super().__init__()
+
+        geometry_format = format.lower()
+        # if inst_has_typename(data, ['ndarray.numpy']):
```
Comment (alberttorosyan) [inline aim/sdk/objects/geometry.py:38]: please remove the commented code

**Review by alberttorosyan — CHANGES_REQUESTED** [review]
looks good! left some comments, mainly related to tracking lists of geometries vs. single geometry objects

**Comment by roubkar** [comment]
Hey @pabhermoso, we have added some changes to this PR, which include:
- Naming refactorings - Geometry => Figure3D, Geometries => Figures3D 
- Instead of tracking a list currently, users can track a 3D figure per step
- UI improvements (styles and code cleanups)

Now these figures will be available only on the Run page, we're planning to add Explorer with a separate PR for future releases.

Thanks for your awesome work!

**Review by roubkar — APPROVED** [review]

**Review by alberttorosyan — APPROVED** [review]
Approving as all the requested changes are addressed.
@pabhermoso, thanks for the contribution! Awesome job 🚀

**Comment by pabhermoso** [comment]
Looks awesome!!! Also, much prefer the Figure3D name. Thanks for the work putting this neatly into the UI :) @roubkar @alberttorosyan

**Pull request merged**